File: server/services/requisitions.js
import { API_CONFIG } from '../config/api.js';
import { getAccessToken } from './auth.js';
import { getAccountDetails } from './accounts.js';

export async function getRequisitions() {
  const token = await getAccessToken();
  const response = await fetch(`${API_CONFIG.baseUrl}/requisitions/`, {
    headers: {
      ...API_CONFIG.headers,
      Authorization: `Bearer ${token}`,
    },
  });

  if (!response.ok) {
    throw new Error(`Failed to fetch requisitions: ${response.status}`);
  }

  const data = await response.json();
  return {
    ...data,
    results: data.results.sort((a, b) => 
      new Date(a.created).getTime() - new Date(b.created).getTime()
    ),
  };
}

export async function getRequisitionDetails(id) {
  const token = await getAccessToken();
  const response = await fetch(`${API_CONFIG.baseUrl}/requisitions/${id}/`, {
    headers: {
      ...API_CONFIG.headers,
      Authorization: `Bearer ${token}`,
    },
  });

  if (!response.ok) {
    throw new Error(`Failed to fetch requisition details: ${response.status}`);
  }

  const requisition = await response.json();
  const accountDetails = await Promise.all(
    requisition.accounts.map(accountId => getAccountDetails(accountId))
  );

  return {
    ...requisition,
    accounts: accountDetails,
  };
}

export async function createNewRequisition(params) {
  const token = await getAccessToken();
  const response = await fetch(`${API_CONFIG.baseUrl}/requisitions/`, {
    method: 'POST',
    headers: {
      ...API_CONFIG.headers,
      Authorization: `Bearer ${token}`,
    },
    body: JSON.stringify({
      institution_id: params.institutionId,
      redirect: params.redirectUrl,
      reference: params.reference,
      user_language: params.userLanguage,
    }),
  });

  if (!response.ok) {
    throw new Error(`Failed to create requisition: ${response.status}`);
  }

  return response.json();
}

export async function removeRequisition(id) {
  const token = await getAccessToken();
  const response = await fetch(`${API_CONFIG.baseUrl}/requisitions/${id}/`, {
    method: 'DELETE',
    headers: {
      ...API_CONFIG.headers,
      Authorization: `Bearer ${token}`,
    },
  });

  if (!response.ok) {
    throw new Error(`Failed to delete requisition: ${response.status}`);
  }
}

File: server/src/services/requisitions.py
import asyncio

import httpx
from datetime import datetime

from .accounts import get_account_details
from .auth import get_access_token, API_CONFIG


async def get_requisitions():
    token = get_access_token()

    async with httpx.AsyncClient() as client:
        response = await client.get(
            f"{API_CONFIG['base_url']}/requisitions/",
            headers={**API_CONFIG["headers"], "Authorization": f"Bearer {token}"},
        )
        response.raise_for_status()
        data = response.json()

        # Sort results by creation date
        data["results"].sort(key=lambda x: datetime.fromisoformat(x["created"]))
        return data


async def get_requisition_details(requisition_id: str):
    token = get_access_token()

    async with httpx.AsyncClient() as client:
        response = await client.get(
            f"{API_CONFIG['base_url']}/requisitions/{requisition_id}/",
            headers={**API_CONFIG["headers"], "Authorization": f"Bearer {token}"},
        )
        response.raise_for_status()
        requisition = response.json()

        # Get details for each account
        account_details = await asyncio.gather(
            *[get_account_details(account_id) for account_id in requisition["accounts"]]
        )

        requisition["accounts"] = account_details
        return requisition


async def create_new_requisition(params: dict):
    token = get_access_token()

    async with httpx.AsyncClient() as client:
        response = await client.post(
            f"{API_CONFIG['base_url']}/requisitions/",
            headers={**API_CONFIG["headers"], "Authorization": f"Bearer {token}"},
            json={
                "institution_id": params["institution_id"],
                "redirect": params["redirect"],
                "reference": params["reference"],
                "user_language": params["user_language"],
            },
        )
        response.raise_for_status()
        return response.json()


async def remove_requisition(requisition_id: str):
    token = get_access_token()

    async with httpx.AsyncClient() as client:
        response = await client.delete(
            f"{API_CONFIG['base_url']}/requisitions/{requisition_id}/",
            headers={**API_CONFIG["headers"], "Authorization": f"Bearer {token}"},
        )
        response.raise_for_status()
